pythonwhat/State.py

```diff
@@ -203,17 +170,58 @@ def parse_int(x):
     def set_active_state(state):
         State.active_state = state
 
-# add property methods for retrieving parser outputs
+# add property methods for retrieving parser outputs --------------------------
 # e.g. state.student_withs
 from functools import partial
-for k, Parser in parser_dict.items():
-    def getx(self, tree_name, Parser=Parser): 
+
+def getx(tree_name, Parser, ext_attr, self): 
+    """getter for Parser outputs"""
+    # return cached output if possible
+    cache_key = tree_name + Parser.__name__
+    if self._parser_cache.get(cache_key):
+        p = self._parser_cache[cache_key]
+    else:
+        # otherwise, run parser over tree
         p = Parser()
         p.visit(getattr(self, tree_name))
-        return p.out
-
-    for s in ['student', 'solution']:
-        setattr(State, s+'_'+k, property(partial(getx, tree_name = s+'_tree')))
-
+        # cache
+        self._parser_cache[cache_key] = p
+    return getattr(p, ext_attr)
+
+def get_func_map(tree_name, ext_attr, self):
+    """getter for FunctionParser outputs, uses pre_exercise_mappings"""
+    cache_key = tree_name + FunctionParser.__name__
+    if self._parser_cache.get(cache_key):
+        p = self._parser_cache[cache_key]
+    else:
+        p = FunctionParser()
+        p.mappings = self.pre_exercise_mappings.copy()
+        p.visit(getattr(self, tree_name))
+        self._parser_cache[cache_key] = p
+    return getattr(p, ext_attr)
+    
+for s in ['student', 'solution']:
```

Comment:
Partials are powerful. I only think they can obscure the code sometimes, IMO. Make sure to explain a bit what's going on. It's not that straightforward always.